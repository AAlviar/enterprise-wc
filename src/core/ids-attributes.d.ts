export const attributes: Record<string, unknown>;
export const prefix: [];
export const version: string;
