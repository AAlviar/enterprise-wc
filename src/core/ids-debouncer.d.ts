/** prevents a function that runs on intervals from running too frequently */
declare function debounce(func: CallableFunction, duration?: number, execAsap?: boolean): void;

export default debounce;
