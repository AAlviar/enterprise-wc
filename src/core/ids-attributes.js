// Please keep constants alphabetized
export const attributes = {
  ACCEPT: 'accept',
  ACTIVE: 'active',
  ADVANCED: 'advanced',
  ALT: 'alt',
  ALTERNATE_ROW_SHADING: 'alternate-row-shading',
  ANIMATED: 'animated',
  ANIMATION_STYLE: 'animation-style',
  ALIGN: 'align',
  ALIGN_EDGE: 'align-edge',
  ALIGN_TARGET: 'align-target',
  ALIGN_X: 'align-x',
  ALIGN_Y: 'align-y',
  ALLOW_LINK: 'allow-link',
  APPEARANCE: 'appearance',
  ARROW: 'arrow',
  ARROW_TARGET: 'arrow-target',
  AUDIBLE: 'audible',
  AUTO: 'auto',
  AUTOGROW: 'autogrow',
  AUTOGROW_MAX_HEIGHT: 'autogrow-max-height',
  AUTO_FIT: 'auto-fit',
  AUTO_HEIGHT: 'auto-height',
  AUTOSELECT: 'autoselect',
  AUTOUPDATE: 'autoupdate',
  AXIS: 'axis',
  BADGE_COLOR: 'badge-color',
  BADGE_POSITION: 'badge-position',
  BG_TRANSPARENT: 'bg-transparent',
  BOLD: 'bold',
  BLEED: 'bleed',
  BTN_TEXT: 'btn-text',
  BUFFER_SIZE: 'buffer-size',
  BUTTONS: 'buttons',
  BUTTON_TEXT: 'button-text',
  CANCEL: 'cancel',
  CAPTURES_FOCUS: 'captures-focus',
  CAPS_LOCK: 'caps-lock',
  CARD_HEIGHT: 'card-height',
  CARD_WIDTH: 'card-width',
  CHAR_MAX_TEXT: 'char-max-text',
  CHAR_REMAINING_TEXT: 'char-remaining-text',
  CHARACTER_COUNTER: 'character-counter',
  CHECKED: 'checked',
  CLEARABLE: 'clearable',
  CLEARABLE_FORCED: 'clearable-forced',
  COL_END: 'col-end',
  COL_SPAN: 'col-span',
  COL_SPAN_XS: 'col-span-xs',
  COL_SPAN_SM: 'col-span-sm',
  COL_SPAN_MD: 'col-span-md',
  COL_SPAN_LG: 'col-span-lg',
  COL_SPAN_XL: 'col-span-xl',
  COL_SPAN_XXL: 'col-span-xxl',
  COL_START: 'col-start',
  COLLAPSE_ICON: 'collapse-icon',
  COLOR: 'color',
  COLOR_VARIANT: 'color-variant',
  COLS: 'cols',
  COMPACT: 'compact',
  COMPLETED_LABEL: 'completed-label',
  COUNT: 'count',
  COPYRIGHT_YEAR: 'copyright-year',
  CLICKABLE: 'clickable',
  CSS_CLASS: 'css-class',
  CURSOR: 'cursor',
  CYCLES_FOCUS: 'cycles-focus',
  DATA: 'data',
  DAY: 'day',
  DELAY: 'delay',
  DELIMITER: 'delimiter',
  DESCRIPTION: 'description',
  DEVICE_SPECS: 'device-specs',
  DIRTY_TRACKER: 'dirty-tracker',
  DISABLE_EVENTS: 'disable-native-events',
  DISABLED: 'disabled',
  DISMISSIBLE: 'dismissible',
  DISPLAY: 'display',
  DRAGGABLE: 'draggable',
  DROPDOWN_ICON: 'dropdown-icon',
  EDGE: 'edge',
  END_DATE: 'end-date',
  END_HOUR: 'end-hour',
  ERROR: 'error',
  EXPAND_ICON: 'expand-icon',
  EXPANDED: 'expanded',
  EXPANDER_TYPE: 'expander-type',
  FALLBACK: 'fallback',
  FIELD: 'field',
  FAVOR: 'favor',
  FIELD_HEIGHT: 'field-height',
  FILE: 'file',
  FILE_NAME: 'file-name',
  FILL: 'fill',
  FIXED: 'fixed',
  FIRST: 'first',
  FIRST_DAY_OF_WEEK: 'first-day-of-week',
  FOCUSABLE: 'focusable',
  FONT_SIZE: 'font-size',
  FONT_WEIGHT: 'font-weight',
  FORMAT: 'format',
  FORMATTER_WIDTH: 'formatter-width',
  GAP: 'gap',
  GAP_X: 'gap-x',
  GAP_Y: 'gap-y',
  GROUP: 'group',
  GROUP_DISABLED: 'group-disabled',
  HANDLE: 'handle',
  HEIGHT: 'height',
  HIDDEN: 'hidden',
  HIDE_DOWN: 'hide-down',
  HIDE_UP: 'hide-up',
  HITBOX: 'hitbox',
  HORIZONTAL: 'horizontal',
  HREF: 'href',
  HEX: 'hex',
  ICON: 'icon',
  ICON_ALIGN: 'icon-align',
  ID: 'id',
  INDETERMINATE: 'indeterminate',
  INLINE: 'inline',
  INITIALS: 'initials',
  IS_DRAGGING: 'is-dragging',
  ITEM_HEIGHT: 'item-height',
  JUSTIFY: 'justify',
  KEEP_OPEN: 'keep-open',
  LABEL: 'label',
  LABEL_AUDIBLE: 'label-audible',
  LABEL_FILETYPE: 'label-filetype',
  LABEL_HIDDEN: 'label-hidden',
  LABEL_PROGRESS: 'label-progress',
  LABEL_REQUIRED: 'label-required',
  LABEL_TOTAL: 'label-total',
  LANGUAGE: 'language',
  LINEAR: 'linear',
  LIST_STYLE: 'list-style',
  LOCALE: 'locale',
  MARKER_SIZE: 'marker-size',
  MARGINS: 'margins',
  MASK: 'mask',
  MASK_GUIDE: 'mask-guide',
  MASK_OPTIONS: 'mask-options',
  MASK_RETAIN_POSITIONS: 'mask-retain-positions',
  MAX: 'max',
  MAX_FILE_SIZE: 'max-file-size',
  MAX_FILES: 'max-files',
  MAX_FILES_IN_PROCESS: 'max-files-in-process',
  MAXLENGTH: 'maxlength',
  MAX_TRANSFORM_X: 'max-transform-x',
  MAX_TRANSFORM_Y: 'max-transform-y',
  MENU: 'menu',
  MESSAGE_TEXT: 'message-text',
  METHOD: 'method',
  MESSAGE: 'message',
  MESSAGE_TITLE: 'message-title',
  MIN: 'min',
  MIN_COL_WIDTH: 'min-col-width',
  MIN_SIZE: 'min-size',
  MIN_TRANSFORM_X: 'min-transform-x',
  MIN_TRANSFORM_Y: 'min-transform-y',
  MINUTE_INTERVAL: 'minute-interval',
  MODE: 'mode',
  MONTH: 'month',
  MORE_ACTIONS: 'more-actions',
  MULTIPLE: 'multiple',
  NAME: 'name',
  NAV_DISABLED: 'nav-disabled',
  NEXT: 'next',
  NO_MARGINS: 'no-margins',
  NO_PADDING: 'no-padding',
  NO_RIPPLE: 'no-ripple',
  NO_TEXT_ELLIPSIS: 'no-text-ellipsis',
  OPACITY: 'opacity',
  OVERFLOW: 'overflow',
  OVERFLOWED: 'overflowed',
  ORIENTATION: 'orientation',
  PAGE_COUNT: 'page-count',
  PAGE_NUMBER: 'page-number',
  PAGE_SIZE: 'page-size',
  PAGINATION: 'pagination',
  PADDING: 'padding',
  PARENT_CONTAINMENT: 'parent-containment',
  PARENT_DISABLED: 'parent-disabled',
  PASSWORD_VISIBLE: 'password-visible',
  PERCENT: 'percent',
  PERCENTAGE_VISIBLE: 'percentage-visible',
  PLACEHOLDER: 'placeholder',
  PLACEMENT: 'placement',
  POSITION_STYLE: 'position-style',
  PARAM_NAME: 'param-name',
  POSITION: 'position',
  PREVIOUS: 'previous',
  PRINTABLE: 'printable',
  PROGRESS_BAR: 'progress-bar',
  PRODUCT_NAME: 'product-name',
  PRODUCT_VERSION: 'product-version',
  LAST: 'last',
  LINK: 'link',
  LINK_TEXT: 'link-text',
  PROGRESS: 'progress',
  PROGRESS_COLOR: 'progress-color',
  READONLY: 'readonly',
  RESIZABLE: 'resizable',
  RESIZE_ON_DRAG_END: 'resize-on-drag-end',
  REQUIRED: 'required',
  REVEAL: 'reveal',
  REVEALABLE_PASSWORD: 'revealable-password',
  ROUND: 'round',
  ROOT_ITEM: 'root-item',
  ROW_END: 'row-end',
  ROW_HEIGHT: 'row-height',
  ROW_SELECTION: 'row-selection',
  ROW_SPAN: 'row-span',
  ROW_START: 'row-start',
  ROWS: 'rows',
  RESET: 'reset',
  RESULT: 'result',
  SAVE_POSITION: 'save-position',
  SCROLLABLE: 'scrollable',
  SCROLL_TOP: 'scroll-top',
  SECOND_INTERVAL: 'second-interval',
  SELECT: 'select',
  SELECTABLE: 'selectable',
  SELECTED: 'selected',
  SELECTION: 'selection',
  SHAPE: 'shape',
  SHOW_BROWSE_LINK: 'show-browse-link',
  SHOW_HORIZONTAL_GRID_LINES: 'show-horizontal-grid-lines',
  SHOW_TIMELINE: 'show-timeline',
  SHOW_TODAY: 'show-today',
  SHOW_VERTICAL_GRID_LINES: 'show-vertical-grid-lines',
  SORTABLE: 'sortable',
  SQUARE: 'square',
  SIZE: 'size',
  SRC: 'src',
  STARS: 'stars',
  START: 'start',
  START_DATE: 'start-date',
  START_HOUR: 'start-hour',
  STATUS: 'status',
  STEP: 'step',
  STEP_NUMBER: 'step-number',
  STEPS_IN_PROGRESS: 'steps-in-progress',
  STICKY: 'sticky',
  SUBMENU: 'submenu',
  SUPRESS_ROW_DEACTIVATION: 'supress-row-deactivation',
  SUPRESS_ROW_DESELECTION: 'supress-row-deselection',
  SWAPPABLE: 'swappable',
  SWATCH: 'swatch',
  SWIPE_TYPE: 'swipe-type',
  TABBABLE: 'tabbable',
  TABINDEX: 'tabindex',
  TARGET: 'target',
  TITLE: 'title',
  TIMEOUT: 'timeout',
  TOOLBAR_TYPE: 'toolbar-type',
  TOOLTIP: 'tooltip',
  TOTAL: 'total',
  TRANSLATE_TEXT: 'translate-text',
  TRIGGER: 'trigger',
  TRIGGERFIELD: 'triggerfield',
  TRIGGER_LABEL: 'trigger-label',
  TEXT: 'text',
  TEXT_ALIGN: 'text-align',
  TEXT_DECORATION: 'text-decoration',
  TEXT_ELLIPSIS: 'text-ellipsis',
  TIMELINE_INTERVAL: 'timeline-interval',
  TOGGLE_COLLAPSE_ICON: 'toggle-collapse-icon',
  TOGGLE_EXPAND_ICON: 'toggle-expand-icon',
  TOGGLE_ICON_ROTATE: 'toggle-icon-rotate',
  TYPE: 'type',
  UNIQUE_ID: 'unique-id',
  URL: 'url',
  USE_DEFAULT_COPYRIGHT: 'use-default-copyright',
  USE_TOGGLE_TARGET: 'use-toggle-target',
  USER_STATUS: 'user-status',
  VALIDATE: 'validate',
  VALIDATION_EVENTS: 'validation-events',
  VALIDATION_HAS_ERROR: 'validation-has-error',
  VALUE: 'value',
  VALUE_SECONDARY: 'value-secondary',
  VERSION: 'version',
  VERTICAL: 'vertical',
  VIEWBOX: 'viewbox',
  VIRTUAL_SCROLL: 'virtual-scroll',
  VISIBLE: 'visible',
  WIDTH: 'width',
  X: 'x',
  XSS_IGNORED_TAGS: 'xss-ignored-tags',
  Y: 'y',
  Y_AXIS_MIN: 'y-axis-min',
  YEAR: 'year'
};
// Please keep constants alphabetized

export const prefix = {
  PREFIX: 'ids'
};

export const breakpoints = {
  xxl: '1440px',
  xl: '1280px',
  lg: '1024px',
  md: '840px',
  sm: '600px',
  xs: '360px'
};

export const version = '0.0.0-beta.17';
