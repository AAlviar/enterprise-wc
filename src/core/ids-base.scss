/* Sass scss styles. Keep Base very small */
@import '../../node_modules/ids-identity/dist/theme-new/tokens/web/theme-new-mixins';
@import '../mixins/sass/ids-general-mixins';

/* Global variables (might want these to become tokens) */
$breakpoint-xxl: 1440px;
$breakpoint-xl: 1280px;
$breakpoint-lg: 1024px;
$breakpoint-md: 840px;
$breakpoint-sm: 600px;
$breakpoint-xs: 360px;

$transition-length: 200ms;
$transition-ease: cubic-bezier(0.4, 0, 0.6, 1);

// Field Sizes
$input-size-xs: 75px;
$input-size-sm: 150px;
$input-size-mm: 225px;
$input-size-md: 300px ;
$input-size-lg: 400px;
$input-size-full: 100%;

// Field Heights
$input-field-height-xs: 28px;
$input-field-height-sm: 30px;
$input-field-height-md: 38px;
$input-field-height-lg: 48px;

// Sometimes we need to "fade" white into transparency, so we need a
// completely-transparent "white" color to use as one end of a gradient.
$transparent-white: rgba(255 255 255 / 0);

// Other semi-transparent colors (temporary until we move these to design tokens)
// See infor-design/design-system#490
$ids-slider-thumb-color-light: rgba(0 114 237 / 0.1);
$ids-slider-thumb-readonly-shadow-color-light: rgba(119 119 124 / 0.1);
$ids-slider-thumb-color-dark: rgba(0 114 237 / 0.1);
$ids-slider-thumb-readonly-shadow-color-dark: rgba(151 151 155 / 0.1);
$ids-slider-thumb-color-contrast: rgba(0 74 153 / 0.1);
$ids-slider-thumb-readonly-shadow-color-contrast: rgba(71 71 76 / 0.1);
$ids-slider-thumb-hover-shadow-light: rgba(0 0 0 / 0.4);
$ids-slider-thumb-hover-shadow-dark: rgba(255 255 255 / 0.4);
$ids-slider-thumb-hover-shadow-contrast: rgba(0 0 0 / 0.4);
