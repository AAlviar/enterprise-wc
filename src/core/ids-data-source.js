import deepClone from '../utils/ids-deep-clone-utils/ids-deep-clone-utils';

/**
 * Handle Attaching Array / Object Data to Components
 * Features (now and future):
 *  - data
 *  - sort
 *  - filter
 *  - read/map/loop
 *  - primaryKey
 *  - retrieval
 *  - CRUD
 *  - paging (pageSize, serverSide, cache)
 *  - aggregates / group by
 *  - events (requestStart, requestEnd, change, error)
 *  - sync (sync back original array)
 */
class IdsDataSource {
  // Holds a reference to the original data
  originalData = [];

  // Holds the data in its current state
  currentData = [];

  /**
   * Sets the data array on the data source object
   * @param {Array | null} value The array to attach
   */
  set data(value) {
    this.currentData = deepClone(value);
    this.originalData = value;
  }

  /**
   * Return the currently used data in its current state
   * @returns {Array | null} The attached array of data in its current state
   */
  get data() { return this.currentData; }

  /**
   * Executes a provided function once for each array element in the current data
   * @param {Function} fn An optional function to iterate the array
   */
  forEach(fn) {
    this.currentData.forEach(fn);
  }

  /**
   * Sort the dataset
   * @param  {string} field The dataset field
   * @param  {boolean} reverse Sort ascending or descending
   * @param  {Function|null} primer Optional primer function
   */
  sort(field, reverse, primer) {
    const sort = this.sortFunction(field, reverse, primer);
    this.currentData.sort(sort);
  }

  /**
   * An overridable array sort function
   * @param  {string} field The dataset field
   * @param  {boolean} reverse Sort ascending or descending
   * @param  {Function} primer Primer function
   * @returns {object} The sorted dataset or it uses
   */
  sortFunction(field, reverse, primer) {
    const key = (x) => (primer ? primer(x[field]) : x[field]);

    return (a, b) => {
      const A = key(a);
      const B = key(b);
      return ((A < B) ? -1 : ((A > B) ? 1 : 0)) * [-1, 1][+!!reverse]; // eslint-disable-line
    };
  }
}

export default IdsDataSource;
