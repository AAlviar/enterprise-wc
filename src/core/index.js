export { attributes, prefix } from './ids-attributes';
export { default as IdsDataSource } from './ids-data-source';
export {
  IdsElement,
  customElement,
  appendIds,
  mix,
  scss
} from './ids-element';
