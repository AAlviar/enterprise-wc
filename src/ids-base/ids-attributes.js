// Please keep constants alphabetized
export const attributes = {
  ACCEPT: 'accept',
  STICKY: 'sticky',
  ALTERNATE_ROW_SHADING: 'alternate-row-shading',
  ANIMATED: 'animated',
  ANIMATION_STYLE: 'animation-style',
  ALIGN: 'align',
  ALIGN_EDGE: 'align-edge',
  ALIGN_TARGET: 'align-target',
  ALIGN_X: 'align-x',
  ALIGN_Y: 'align-y',
  APPEARANCE: 'appearance',
  ARROW: 'arrow',
  ARROW_TARGET: 'arrow-target',
  AUDIBLE: 'audible',
  AUTO: 'auto',
  AUTOGROW: 'autogrow',
  AUTOGROW_MAX_HEIGHT: 'autogrow-max-height',
  AUTO_HEIGHT: 'auto-height',
  AUTOSELECT: 'autoselect',
  BG_TRANSPARENT: 'bg-transparent',
  BLEED: 'bleed',
  BUTTONS: 'buttons',
  CANCEL: 'cancel',
  CHAR_MAX_TEXT: 'char-max-text',
  CHAR_REMAINING_TEXT: 'char-remaining-text',
  CHARACTER_COUNTER: 'character-counter',
  CHECKED: 'checked',
  CLEARABLE: 'clearable',
  CLEARABLE_FORCED: 'clearable-forced',
  COL_END: 'col-end',
  COL_SPAN: 'col-span',
  COL_START: 'col-start',
  COLOR: 'color',
  COLS: 'cols',
  COMPACT: 'compact',
  COUNT: 'count',
  CLICKABLE: 'clickable',
  CSS_CLASS: 'css-class',
  DATA: 'data',
  DELAY: 'delay',
  DIRTY_TRACKER: 'dirty-tracker',
  DISABLE_EVENTS: 'disable-native-events',
  DISABLED: 'disabled',
  DISMISSIBLE: 'dismissible',
  DISPLAY: 'display',
  ERROR: 'error',
  EXPANDED: 'expanded',
  FIELD_HEIGHT: 'field-height',
  FILE: 'file',
  FILE_NAME: 'file-name',
  FILL: 'fill',
  FIXED: 'fixed',
  FIRST: 'first',
  FOCUSABLE: 'focusable',
  FONT_SIZE: 'font-size',
  FONT_WEIGHT: 'font-weight',
  GAP: 'gap',
  GROUP: 'group',
  GROUP_DISABLED: 'group-disabled',
  HORIZONTAL: 'horizontal',
  HREF: 'href',
  HEX: 'hex',
  ICON: 'icon',
  ICON_ALIGN: 'icon-align',
  ID: 'id',
  INDETERMINATE: 'indeterminate',
  INLINE: 'inline',
  JUSTIFY: 'justify',
  KEEP_OPEN: 'keep-open',
  LABEL: 'label',
  LABEL_AUDIBLE: 'label-audible',
  LABEL_HIDDEN: 'label-hidden',
  LABEL_REQUIRED: 'label-required',
  LABEL_FILETYPE: 'label-filetype',
  LINEAR: 'linear',
  MAX: 'max',
  MAX_FILE_SIZE: 'max-file-size',
  MAX_FILES: 'max-files',
  MAX_FILES_IN_PROCESS: 'max-files-in-process',
  MASK: 'mask',
  MASK_GUIDE: 'mask-guide',
  MASK_OPTIONS: 'mask-options',
  MASK_RETAIN_POSITIONS: 'mask-retain-positions',
  MAXLENGTH: 'maxlength',
  MENU: 'menu',
  METHOD: 'method',
  MESSAGE: 'message',
  MESSAGE_TITLE: 'message-title',
  MIN: 'min',
  MODE: 'mode',
  MULTIPLE: 'multiple',
  NAME: 'name',
  NAV_DISABLED: 'nav-disabled',
  NEXT: 'next',
  NO_MARGINS: 'no-margins',
  NO_RIPPLE: 'no-ripple',
  NO_TEXT_ELLIPSIS: 'no-text-ellipsis',
  OPACITY: 'opacity',
  OVERFLOW: 'overflow',
  ORIENTATION: 'orientation',
  PAGE_COUNT: 'page-count',
  PAGE_NUMBER: 'page-number',
  PAGE_SIZE: 'page-size',
  PADDING: 'padding',
  PARENT_DISABLED: 'parent-disabled',
  PERCENTAGE_VISIBLE: 'percentage-visible',
  PLACEHOLDER: 'placeholder',
  PLACEMENT: 'placement',
  POSITION_STYLE: 'position-style',
  PARAM_NAME: 'param-name',
  PREVIOUS: 'previous',
  PRINTABLE: 'printable',
  LAST: 'last',
  PROGRESS: 'progress',
  READONLY: 'readonly',
  RESIZABLE: 'resizable',
  REQUIRED: 'required',
  ROW_END: 'row-end',
  ROW_HEIGHT: 'row-height',
  ROW_SPAN: 'row-span',
  ROW_START: 'row-start',
  ROWS: 'rows',
  SCROLLABLE: 'scrollable',
  SELECT: 'select',
  SELECTED: 'selected',
  SHAPE: 'shape',
  SHOW_BROWSE_LINK: 'show-browse-link',
  SQUARE: 'square',
  SIZE: 'size',
  START: 'start',
  STATUS: 'status',
  STEP: 'step',
  STEP_NUMBER: 'step-number',
  SUBMENU: 'submenu',
  SWIPE_TYPE: 'swipe-type',
  TABBABLE: 'tabbable',
  TABINDEX: 'tabindex',
  TARGET: 'target',
  TITLE: 'title',
  TOOLTIP: 'tooltip',
  TOTAL: 'total',
  TRIGGER: 'trigger',
  TRIGGERFIELD: 'triggerfield',
  TRIGGER_LABEL: 'trigger-label',
  TEXT: 'text',
  TEXT_ALIGN: 'text-align',
  TEXT_DECORATION: 'text-decoration',
  TEXT_ELLIPSIS: 'text-ellipsis',
  TYPE: 'type',
  URL: 'url',
  VALIDATE: 'validate',
  VALIDATION_EVENTS: 'validation-events',
  VALIDATION_HAS_ERROR: 'validation-has-error',
  VALUE: 'value',
  VERSION: 'version',
  VERTICAL: 'vertical',
  VIRTUAL_SCROLL: 'virtual-scroll',
  VISIBLE: 'visible',
  X: 'x',
  XSS_IGNORED_TAGS: 'xss-ignored-tags',
  Y: 'y'
};
// Please keep constants alphabetized

export const prefix = {
  PREFIX: 'ids'
};
