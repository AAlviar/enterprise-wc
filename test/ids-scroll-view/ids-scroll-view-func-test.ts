/**
 * @jest-environment jsdom
 */
import { attributes } from '../../src/core/ids-attributes';
import wait from '../helpers/wait';
import IdsScrollView from '../../src/components/ids-scroll-view/ids-scroll-view';
import IntersectionObserver from '../helpers/intersection-observer-mock';

describe('IdsScrollView Component', () => {
  let scrollView: any;
  const html = `<img slot="scroll-view-item" src="../assets/images/camera-1.png" alt="Slide 1, Sony Camera, Front"/>
  <img slot="scroll-view-item" src="../assets/images/camera-2.png" alt="Slide 3, Sony Camera, Back Display"/>
  <img slot="scroll-view-item" src="../assets/images/camera-3.png" alt="Slide 3, Sony Camera, From Top"/>
  <img slot="scroll-view-item" src="../assets/images/camera-4.png" alt="Slide 4, Olympus Camera, Front"/>
  <img slot="scroll-view-item" src="../assets/images/camera-5.png" alt="Slide 5, Olympus Camera, Exposed to water"/>
  <img slot="scroll-view-item" src="../assets/images/camera-6.png" alt="Slide 6, Sony E-mount Camera, Front"/>`;

  beforeEach(async () => {
    // Mock IntersectionObserver
    (<any>window).IntersectionObserver = IntersectionObserver;

    // Append the element
    const elem: any = new IdsScrollView();
    elem.innerHTML = html;
    document.body.appendChild(elem);
    scrollView = document.querySelector('ids-scroll-view');

    // Mock scrollBy
    elem.container.scrollBy = jest.fn();
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders with no errors', () => {
    scrollView.remove();
    const errors = jest.spyOn(global.console, 'error');
    const elem: any = new IdsScrollView();
    elem.innerHTML = html;
    document.body.appendChild(elem);
    expect(document.querySelectorAll('ids-scroll-view').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
    elem.remove();
  });

  it('renders correctly', () => {
    expect(scrollView.shadowRoot.innerHTML).toMatchSnapshot();
  });

  it('should sets auto play', () => {
    expect(scrollView.getAttribute(attributes.AUTO_PLAY)).toEqual(null);
    expect(scrollView.autoPlay).toEqual(false);
    scrollView.autoPlay = true;
    expect(scrollView.getAttribute(attributes.AUTO_PLAY)).toEqual('');
    expect(scrollView.autoPlay).toEqual(true);
    scrollView.autoPlay = false;
    expect(scrollView.getAttribute(attributes.AUTO_PLAY)).toEqual(null);
    expect(scrollView.autoPlay).toEqual(false);
  });

  it('should sets delay', () => {
    const defaultVal = 3000;
    expect(scrollView.getAttribute(attributes.DELAY)).toEqual(null);
    expect(scrollView.delay).toEqual(defaultVal);
    scrollView.delay = 2000;
    expect(scrollView.getAttribute(attributes.DELAY)).toEqual('2000');
    expect(scrollView.delay).toEqual(2000);
    scrollView.delay = 'test';
    expect(scrollView.getAttribute(attributes.DELAY)).toEqual(null);
    expect(scrollView.delay).toEqual(defaultVal);
  });

  it('should sets loop', () => {
    expect(scrollView.getAttribute(attributes.LOOP)).toEqual(null);
    expect(scrollView.loop).toEqual(false);
    scrollView.loop = true;
    expect(scrollView.getAttribute(attributes.LOOP)).toEqual('');
    expect(scrollView.loop).toEqual(true);
    scrollView.loop = false;
    expect(scrollView.getAttribute(attributes.LOOP)).toEqual(null);
    expect(scrollView.loop).toEqual(false);
  });

  it('should sets loop reverse', () => {
    expect(scrollView.getAttribute(attributes.LOOP_REVERSE)).toEqual(null);
    expect(scrollView.loopReverse).toEqual(false);
    scrollView.loopReverse = true;
    expect(scrollView.getAttribute(attributes.LOOP_REVERSE)).toEqual('');
    expect(scrollView.loopReverse).toEqual(true);
    scrollView.loopReverse = false;
    expect(scrollView.getAttribute(attributes.LOOP_REVERSE)).toEqual(null);
    expect(scrollView.loopReverse).toEqual(false);
  });

  it('should sets show tooltip', () => {
    expect(scrollView.getAttribute(attributes.SHOW_TOOLTIP)).toEqual(null);
    expect(scrollView.showTooltip).toEqual(false);
    scrollView.showTooltip = true;
    expect(scrollView.getAttribute(attributes.SHOW_TOOLTIP)).toEqual('');
    expect(scrollView.showTooltip).toEqual(true);
    scrollView.showTooltip = false;
    expect(scrollView.getAttribute(attributes.SHOW_TOOLTIP)).toEqual(null);
    expect(scrollView.showTooltip).toEqual(false);
  });

  it('should sets suppress controls', () => {
    expect(scrollView.getAttribute(attributes.SUPPRESS_CONTROLS)).toEqual(null);
    expect(scrollView.suppressControls).toEqual(false);
    scrollView.suppressControls = true;
    expect(scrollView.getAttribute(attributes.SUPPRESS_CONTROLS)).toEqual('');
    expect(scrollView.suppressControls).toEqual(true);
    scrollView.suppressControls = false;
    expect(scrollView.getAttribute(attributes.SUPPRESS_CONTROLS)).toEqual(null);
    expect(scrollView.suppressControls).toEqual(false);
  });

  it('should call control api', async () => {
    const delay = 10;
    scrollView.delay = delay;

    const slideWait = async (num: number) => {
      await wait((delay * num) + 1);
    };

    const slideExpect = async (currentNum: number, waitNum: number, expectNum: number) => {
      await slideWait(currentNum + waitNum);
      const selEl = scrollView.controls.querySelector('.selected');
      expect(selEl.getAttribute('data-slide-number')).toEqual(`${expectNum}`);
    };

    scrollView.autoPlay = true;
    await slideExpect(0, 2, 2);

    scrollView.stop();
    await slideExpect(2, 2, 2);

    scrollView.play();
    await slideExpect(2, 3, 5);

    scrollView.toggle();
    await slideExpect(5, 3, 5);

    scrollView.first();
    await slideExpect(5, 1, 0);

    scrollView.next();
    await slideExpect(0, 1, 1);

    scrollView.previous();
    await slideExpect(1, 1, 0);

    scrollView.last();
    await slideExpect(0, 1, 5);

    scrollView.slideTo(2);
    await slideExpect(5, 1, 2);

    scrollView.loopReverse = true;
    scrollView.toggle();
    await slideExpect(2, 2, 3);

    scrollView.autoPlay = false;
    scrollView.play();
    await slideExpect(2, 2, 3);
  });

  it('can click the circle buttons', () => {
    const link = scrollView.controls.querySelector('[data-slide-number="4"]');
    expect(link.classList.contains('selected')).toBeFalsy();
    expect(link.getAttribute('aria-selected')).toEqual(null);

    link.click();
    expect(link.classList.contains('selected')).toBeTruthy();
    expect(link.getAttribute('aria-selected')).toEqual('true');
  });

  it.skip('can click the control area and nothing happens', () => {
    scrollView.shadowRoot.querySelector('.ids-scroll-view-controls').click();
    const link = scrollView.controls.querySelector('[data-slide-number="0"]');
    expect(link.classList.contains('selected')).toBeTruthy();
    expect(link.getAttribute('aria-selected')).toEqual('true');
  });

  it('moved on ArrowLeft/ArrowRight', () => {
    const testArrowKey = (key: string, id: string) => {
      scrollView.dispatchEvent(new KeyboardEvent('keydown', { key }));
      const link = scrollView.controls.querySelector(`[data-slide-number="${id}"]`);
      expect(link.classList.contains('selected')).toBeTruthy();
      expect(link.getAttribute('aria-selected')).toEqual('true');
    };

    // One back wont do anything
    testArrowKey('ArrowLeft', '0');
    // key all the way right
    testArrowKey('ArrowRight', '1');
    testArrowKey('ArrowRight', '2');
    testArrowKey('ArrowRight', '3');
    testArrowKey('ArrowRight', '4');
    testArrowKey('ArrowRight', '5');
    // One extra wont do anything
    testArrowKey('ArrowRight', '5');
    // key all the way back
    testArrowKey('ArrowLeft', '4');
    testArrowKey('ArrowLeft', '3');
    testArrowKey('ArrowLeft', '2');
    testArrowKey('ArrowLeft', '1');
    testArrowKey('ArrowLeft', '0');
    testArrowKey('ArrowLeft', '0');
    testArrowKey('Enter', '0');
  });
});
