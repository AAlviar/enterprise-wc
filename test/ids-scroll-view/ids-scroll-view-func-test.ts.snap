// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsScrollView Component renders correctly 1`] = `
"<style nonce=\\"0a59a005\\">:host { background-color: transparent; }</style><div class=\\"ids-scroll-view-container\\" part=\\"container\\">
        <div class=\\"ids-scroll-view\\" part=\\"scroll-view\\" role=\\"complementary\\" tabindex=\\"-1\\">
          <slot name=\\"scroll-view-item\\"></slot>
        </div>
        <div class=\\"ids-scroll-view-controls\\" part=\\"controls\\" role=\\"tablist\\">
        <a class=\\"selected\\" href=\\"#id-0\\" part=\\"button\\" tabindex=\\"0\\" role=\\"tab\\" aria-selected=\\"true\\"><span class=\\"audible\\">Slide 1, Sony Camera, Front</span></a><a href=\\"#id-1\\" part=\\"button\\" tabindex=\\"-1\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"audible\\">Slide 3, Sony Camera, Back Display</span></a><a href=\\"#id-2\\" part=\\"button\\" tabindex=\\"-1\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"audible\\">Slide 3, Sony Camera, From Top</span></a><a href=\\"#id-3\\" part=\\"button\\" tabindex=\\"-1\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"audible\\">Slide 4, Olympus Camera, Front</span></a><a href=\\"#id-4\\" part=\\"button\\" tabindex=\\"-1\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"audible\\">Slide 5, Olympus Camera, Exposed to water</span></a><a href=\\"#id-5\\" part=\\"button\\" tabindex=\\"-1\\" role=\\"tab\\" aria-selected=\\"false\\"><span class=\\"audible\\">Slide 6, Sony E-mount Camera, Front</span></a></div>
    </div>"
`;
